frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -0,0 +1,839 @@
+			MinerError::PreDispatchChecksFailed
+		})?;
+
+		// submit the raw solution to the pool.
+		let call = Call::submit_unsigned(raw_solution, witness).into();
+
+		SubmitTransaction::<T, Call<T>>::submit_unsigned_transaction(call)
+			.map_err(|_| MinerError::PoolSubmissionFailed)
+	}
+
+	/// Mine a new npos solution.
+	pub fn mine_solution(
+		iters: usize,
+	) -> Result<(RawSolution<CompactOf<T>>, SolutionOrSnapshotSize), MinerError> {
+		let RoundSnapshot { voters, targets } =
+			Self::snapshot().ok_or(MinerError::SnapshotUnAvailable)?;
+		let desired_targets = Self::desired_targets().ok_or(MinerError::SnapshotUnAvailable)?;
+
+		seq_phragmen::<_, CompactAccuracyOf<T>>(
+			desired_targets as usize,
+			targets,
+			voters,
+			Some((iters, 0)),
+		)
+		.map_err(Into::into)
+		.and_then(Self::prepare_election_result)
+	}
+
+	/// Convert a raw solution from [`sp_npos_elections::ElectionResult`] to [`RawSolution`], which
+	/// is ready to be submitted to the chain.
+	///
+	/// Will always reduce the solution as well.
+	pub fn prepare_election_result(
+		election_result: ElectionResult<T::AccountId, CompactAccuracyOf<T>>,
+	) -> Result<(RawSolution<CompactOf<T>>, SolutionOrSnapshotSize), MinerError> {
+		// NOTE: This code path is generally not optimized as it is run offchain. Could use some at
+		// some point though.
+
+		// storage items. Note: we have already read this from storage, they must be in cache.
+		let RoundSnapshot { voters, targets } =
+			Self::snapshot().ok_or(MinerError::SnapshotUnAvailable)?;
+		let desired_targets = Self::desired_targets().ok_or(MinerError::SnapshotUnAvailable)?;
+
+		// closures.
+		let cache = helpers::generate_voter_cache::<T>(&voters);
+		let voter_index = helpers::voter_index_fn::<T>(&cache);
+		let target_index = helpers::target_index_fn_linear::<T>(&targets);
+		let voter_at = helpers::voter_at_fn::<T>(&voters);
+		let target_at = helpers::target_at_fn::<T>(&targets);
+		let stake_of = helpers::stake_of_fn::<T>(&voters, &cache);
+
+		let ElectionResult { assignments, winners } = election_result;
+
+		// convert to staked and reduce.
+		let mut staked = assignment_ratio_to_staked_normalized(assignments, &stake_of)
+			.map_err::<MinerError, _>(Into::into)?;
+		sp_npos_elections::reduce(&mut staked);
+
+		// convert back to ration and make compact.
+		let ratio = assignment_staked_to_ratio_normalized(staked)?;
+		let compact = <CompactOf<T>>::from_assignment(ratio, &voter_index, &target_index)?;
+
+		let size =
+			SolutionOrSnapshotSize { voters: voters.len() as u32, targets: targets.len() as u32 };
+		let maximum_allowed_voters = Self::maximum_voter_for_weight::<T::WeightInfo>(
+			desired_targets,
+			size,
+			T::MinerMaxWeight::get(),
+		);
+		log!(
+			debug,
+			"miner: current compact solution voters = {}, maximum_allowed = {}",
+			compact.voter_count(),
+			maximum_allowed_voters,
+		);
+		let compact = Self::trim_compact(maximum_allowed_voters, compact, &voter_index)?;
```

Comment:
it is not possible that trim removes a winner, isn't it ?

Reply:
In theory, it is possible, and if so, because the solution is submitted using `pub(crate) fn mine_and_submit()`, and there we do the `unsigned_pre_dispatch_checks`, which checks that the number of winners is correct.

About it happening: a winner is removed from the set if they are only backed by the edges that get removed. Given that we only remove the smallest of edges, I see this as highly unlikely, and there's nothing more that we can do about it, as if we don't remove the small edges, the solution won't be accepted anyhow. 

Finally, if this happens, validators will all fail to submit and we do on-chain. 

Quite a detail that you pointed out, but I don't think we can do anything about it for now.

Reply:
ah I see we check before submitting, but better to have it noted in the doc that it can happen